### PR TITLE
[SPARK-24626] [SQL] Improve location size calculation in Analyze Table command

### PR DESCRIPTION
**Pull request by Achuth17**
## What changes were proposed in this pull request?

Currently, Analyze table calculates table size sequentially for each partition. We can parallelize size calculations over partitions.

Results : Tested on a table with 100 partitions and data stored in S3. 
With changes :
- 10.429s
- 10.557s
- 10.439s
- 9.893s 

Without changes :
- 110.034s
- 99.510s
- 100.743s
- 99.106s

## How was this patch tested?

Simple unit test.

### CODE REVIEW

**Comment by maropu** [comment]
This pr improves actual performance values? (My question is that the calculation is a bottleneck?)

**Comment by Achuth17** [comment]
Yes, In the case where the data is stored in S3 I noticed a significant difference.

Some rough numbers - When done serially for a table in S3 with 1000 partitions, the calculateTotalSize method took about 90 seconds vs 30-40 seconds when done in parallel.

**Comment by maropu** [comment]
cc: @wzhfy @gatorsmile

**Comment by gatorsmile** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 1]
```diff
@@ -47,22 +47,34 @@ object CommandUtils extends Logging {
...
+          catalogTable.storage.locationUri, stagingDir)
     } else {
       // Calculate table size as a sum of the visible partitions. See SPARK-21079
       val partitions = sessionState.catalog.listPartitions(catalogTable.identifier)
-      partitions.map { p =>
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val numParallelism = Math.min(partitions.size,
+        Math.min(spark.sparkContext.defaultParallelism, 10000))
+      spark.sparkContext.parallelize(partitions, numParallelism).mapPartitions {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: The direction is right, but the implementation needs to be improved by using `InMemoryFileIndex.bulkListLeafFiles`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 2]
```diff
@@ -47,15 +48,26 @@ object CommandUtils extends Logging {
...
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val paths = partitions.map(x => new Path(x.storage.locationUri.get.getPath))
+      val pathFilter = new PathFilter {
+        override def accept(path: Path): Boolean = {
+          !path.getName.startsWith(stagingDir)
+        }
+      }
+      val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(paths,
+        sessionState.newHadoopConf(), pathFilter, spark).flatMap(x => x._2)
```
Comment (Achuth17) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: The tests are passing but this line is incorrect. 

@gatorsmile @maropu, PathFilter is not serializable, How do we pass PathFilter in a serializable manner? I checked the code and one other way is to use `FileInputFormat.getInputPathFilter`/`FileInputFormat.setInputPathFilter` but I couldn't get it to work. 

Also, is it okay if we filter the `Seq[(Path, Seq[FileStatus])]` returned by `bulkListLeafFiles` and remove `stagingDir` files?

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: The above approach might not work too. In the earlier implementation there was a check from recursively listing files from certain directories (`stagingDir`) and having a pathFilter might not be the right approach. 

So I wanted to introduce a list of strings called `filterDir` as a new parameter to `bulkListLeafFiles` which can be used to check if a particular directory can be recursed further.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]:
```Scala
class PathFilterIgnoreNonData(stagingDir: String) extends PathFilter with Serializable {
  override def accept(path: Path): Boolean = {
    val fileName = path.getName
    (!fileName.startsWith(stagingDir) &&
      // Ignore metadata files starting with "_" (for example, files created by
      // DirectoryAtomicCommitProtocol) when computing the location size
      !fileName.startsWith("_"))
  }
}
```

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: Thank you, I have made the changes. Can you review this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 3]
```diff
@@ -47,15 +48,29 @@ object CommandUtils extends Logging {
...
     if (catalogTable.partitionColumnNames.isEmpty) {
       calculateLocationSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
     } else {
       // Calculate table size as a sum of the visible partitions. See SPARK-21079
       val partitions = sessionState.catalog.listPartitions(catalogTable.identifier)
-      partitions.map { p =>
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val paths = partitions.map(x => new Path(x.storage.locationUri.get))
+      val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: Since `SerializablePathFilter` is only used here, how about defining it as an anonymous class?
```
      val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
      val pathFilter = new PathFilter with Serializable {
        override def accept(path: Path): Boolean = ...
      }
      val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(paths,
        sessionState.newHadoopConf(), pathFilter,
        spark).flatMap(_._2)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 4]
```diff
@@ -162,7 +162,7 @@ object InMemoryFileIndex extends Logging {
    *
    * @return for each input path, the set of discovered files for the path
    */
-  private def bulkListLeafFiles(
+   def bulkListLeafFiles(
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Remove the unnecessary indent. Also, `private[sql]`?

**Comment by maropu** [comment]
Can you add tests by referring the existing tests, e.g., `FileIndexSuite`?

**Comment by maropu** [comment]
Super nit though, I think you'd be better to put the performance values in the section `What changes were proposed in this pull request?` instead of the new `Results` section in the pr description.

**Comment by Achuth17** [comment]
@maropu regarding tests, since I didn't make any changes to `inMemoryFileIndex` isn't it better if I added tests in StatisticsCollectionSuite/StatisticsSuite? 

Also, since we just improved an existing method how do we test this?

**Comment by maropu** [comment]
I think it's ok to put tests in `CommandUtilsSuite`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala (hunk 3 above)
Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 4 above)
Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Done.

**Comment by Achuth17** [comment]
@maropu, The method `calculateTotalSize` is already tested in `StatisticsSuite` in a few places. One such test is "analyze Hive serde tables" where we check to make sure the calculated size of the table  is the same as the expected value after analyze table command. Should we add more tests?

**Comment by maropu** [comment]
you'd be better to do that. cc: @gatorsmile

**Comment by Achuth17** [comment]
@maropu I have added a simple test, Can you check it out?

**Comment by Achuth17** [comment]
Ping @gatorsmile

**Comment by holdenk** [comment]
Maybe @ericl has some cycles?

**Comment by Achuth17** [comment]
@maropu @gatorsmile Can you review this? Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 5]
```diff
@@ -148,6 +148,19 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
     }
   }
 
+  test("verify table size calculation is accurate") {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: I originally meant you'd be better to test if the calculation done in parallel like https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala#L99

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 6]
```diff
@@ -47,15 +48,27 @@ object CommandUtils extends Logging {
...
+      val pathFilter = new PathFilter with Serializable {
+        override def accept(path: Path): Boolean = {
+          val fileName = path.getName
+          (!fileName.startsWith(stagingDir) &&
+            // Ignore metadata files starting with "_"
+            !fileName.startsWith("_"))
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: We already have the function to check if it is a data file, or not. Can we reuse this? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala#L232

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: So we are not creating an instance of `InMemoryFileIndex` class so `isDataPath` is not accessible here. I think it would be simpler to add these checks here.

**Comment by Achuth17** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala (hunk 6 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: How about just moving the function into somewhere that can be accessed from here?

**Comment by holdenk** [comment]
Jenkins retest this please.

**Comment by gatorsmile** [comment]
ping me next week. I will review this again. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala (hunk 6 above)
Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: It is a generic method and only used within `PartitioningAwareFileIndex`, so we can move it to `CommandUtils`. Does that sound okay?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: How about `DataSourceUtils`?

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: Done. 
Also, we are not doing this check when `calculateLocationSize` is called directly. I will file a different PR for this as this is not related to AnalyzeTableCommand.

**Comment by Achuth17** [comment]
@gatorsmile @maropu ping.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala [hunk 7]
```diff
@@ -55,4 +57,11 @@ private[sql] object DataSourceV2Utils extends Logging {
 
     case _ => Map.empty
   }
+
+  // SPARK-15895: Metadata files (e.g. Parquet summary files) and temporary files should not be
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala]: Why do you use `DataSourceV2Utils` instead of `DataSourceUtils`?

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala]: Moved it, I couldn't see that file before pulling the upstream master.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 5 above)
Comment (Achuth17) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: @maropu, Does this test look okay?

**Comment by Achuth17** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 8]
```diff
@@ -47,15 +48,26 @@ object CommandUtils extends Logging {
...
-      partitions.map { p =>
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val paths = partitions.map(x => new Path(x.storage.locationUri.get))
+      val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
+      val pathFilter = new PathFilter with Serializable {
+        override def accept(path: Path): Boolean = {
+          val fileName = path.getName
+          (// Ignore metadata files starting with "_"
+            DataSourceUtils.isDataPath(path) && !fileName.startsWith(stagingDir))
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: How about this?
```
override def accept(path: Path): Boolean = {
  DataSourceUtils.isDataPath(path) && !path.getName.startsWith(stagingDir))
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 9]
```diff
@@ -47,15 +48,26 @@ object CommandUtils extends Logging {
...
+        }
+      }
+      val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(paths,
+        sessionState.newHadoopConf(), pathFilter,
+        spark).flatMap(_._2)
+      fileStatusSeq.map(_.getLen).sum
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: How about this?
```
      val fileStatusSeq = InMemoryFileIndex.bulkListLeafFiles(
        paths, sessionState.newHadoopConf(), SerializablePathFilter(stagingDir), spark)
      fileStatusSeq.flatMap(_._2.map(_.getLen)).sum
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 10]
```diff
@@ -59,14 +59,15 @@ abstract class PartitioningAwareFileIndex(
   override def listFiles(
       partitionFilters: Seq[Expression], dataFilters: Seq[Expression]): Seq[PartitionDirectory] = {
     val selectedPartitions = if (partitionSpec().partitionColumns.isEmpty) {
-      PartitionDirectory(InternalRow.empty, allFiles().filter(f => isDataPath(f.getPath))) :: Nil
+      PartitionDirectory(InternalRow.empty,
+        allFiles().filter(f => DataSourceUtils.isDataPath(f.getPath))) :: Nil
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: super nit: 
```
      val files = allFiles().filter(f => DataSourceUtils.isDataPath(f.getPath))
      PartitionDirectory(InternalRow.empty, files) :: Nil
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 5 above)
Comment (maropu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: We should make the title more obvious and how about `SPARK-24626 parallelize location size calculation in Analyze Table command`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 11]
```diff
@@ -148,6 +149,25 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+    val checkSizeTable = "checkSizeTable"
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: super nit: can you move this variable inside `withSQLConf`?

**Comment by maropu** [comment]
LGTM except for minor comments

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 11 above)
Comment (Achuth17) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Done.

**Comment by Achuth17** [comment]
@gatorsmile Can you review this please? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 12]
```diff
@@ -47,15 +48,23 @@ object CommandUtils extends Logging {
     }
   }
 
-  def calculateTotalSize(sessionState: SessionState, catalogTable: CatalogTable): BigInt = {
+  def calculateTotalSize(spark: SparkSession, catalogTable: CatalogTable): BigInt = {
+    val sessionState = spark.sessionState
     if (catalogTable.partitionColumnNames.isEmpty) {
       calculateLocationSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: It sounds like we are not following the same file filtering for non-partition tables.

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: This filtering of temp files is not done in other places where calculateLocationSize is called directly (eg AlterTableAddPartition). Should we skip checking for the temp files since they don't make a lot of difference to the total size?

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: If not, we have to make changes in all those places too.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: Let us make them consistent. only count the data files.

**Comment by gatorsmile** [comment]
Also make it configurable. At least, users can turn this feature off.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala (hunk 12 above)
Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: Got it, maybe we can add the filter inside calculateLocationSize. Let me know if that's okay, I can make the changes.

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: I have made the changes for this in calculateLocationSize.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 13]
```diff
@@ -47,15 +48,23 @@ object CommandUtils extends Logging {
...
       // Calculate table size as a sum of the visible partitions. See SPARK-21079
       val partitions = sessionState.catalog.listPartitions(catalogTable.identifier)
-      partitions.map { p =>
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val paths = partitions.map(x => new Path(x.storage.locationUri.get))
+      val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
+      val pathFilter = new PathFilter with Serializable {
+        override def accept(path: Path): Boolean = {
+          DataSourceUtils.isDataPath(path) && !path.getName.startsWith(stagingDir)
```
[the remark this answers is not in the document]
Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: This is a private method in partitioningAwareFileIndex, Should I add this change to the migration guide?

**Comment by Achuth17** [comment]
@gatorsmile, I have made the changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 14]
```diff
@@ -1449,6 +1449,13 @@ object SQLConf {
     .intConf
     .checkValues((1 to 9).toSet + Deflater.DEFAULT_COMPRESSION)
     .createWithDefault(Deflater.DEFAULT_COMPRESSION)
+
+  val COMPUTE_STATS_LIST_FILES_IN_PARALLEL =
+    buildConf("spark.sql.execution.computeStatsListFilesInParallel")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: How about `spark.sql.parallelFileListingInCommands.enabled`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 15]
```diff
@@ -1449,6 +1449,13 @@ object SQLConf {
...
+      .internal()
+      .doc("If True, File listing for compute statistics is done in parallel.")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: When true, SQL commands use parallel file listing, as opposed to single thread listing. This usually speeds up commands that need to list many directories.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 16]
```diff
@@ -78,7 +93,8 @@ object CommandUtils extends Logging {
       val size = if (fileStatus.isDirectory) {
         fs.listStatus(path)
           .map { status =>
-            if (!status.getPath.getName.startsWith(stagingDir)) {
+            if (!status.getPath.getName.startsWith(stagingDir) &&
+              DataSourceUtils.isDataPath(path)) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala]: This is also a behavior change. Could you document it too?

File: docs/sql-programming-guide.md [hunk 17]
```diff
@@ -1872,6 +1872,8 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
   - In version 2.3 and earlier, CSV rows are considered as malformed if at least one column value in the row is malformed. CSV parser dropped such rows in the DROPMALFORMED mode or outputs an error in the FAILFAST mode. Since Spark 2.4, CSV row is considered as malformed only when it contains malformed column values requested from CSV datasource, other values can be ignored. As an example, CSV file contains the "id,name" header and one row "1234". In Spark 2.4, selection of the id column consists of a row with one column value 1234 but in Spark 2.3 and earlier it is empty in the DROPMALFORMED mode. To restore the previous behavior, set `spark.sql.csv.parser.columnPruning.enabled` to `false`.
+  - Since Spark 2.4, File listing for compute statistics is done in parallel by default. This can be disabled by setting `spark.sql.execution.computeStatsListFilesInParallel` to `False`.
```
Comment (maropu) [inline docs/sql-programming-guide.md]: nit: `File listing for compute statistics in the ANALYZE command...`?

File: docs/sql-programming-guide.md [hunk 18]
```diff
@@ -1872,6 +1872,8 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
+  - Since Spark 2.4, Metadata files (e.g. Parquet summary files) and temporary files are not counted as data files when calculating table size during Statistics computation.
```
Comment (maropu) [inline docs/sql-programming-guide.md]: nit: `... during Statistics computation in the ANALYZE command.`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 19]
```diff
@@ -1449,6 +1449,15 @@ object SQLConf {
     .intConf
     .checkValues((1 to 9).toSet + Deflater.DEFAULT_COMPRESSION)
     .createWithDefault(Deflater.DEFAULT_COMPRESSION)
+
+  val PARALLEL_FILE_LISTING_IN_COMMMANDS =
+    buildConf("spark.sql.parallelFileListingInCommands.enabled")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Currently, in the master, this listing happens only in the ANALYZE SQL command, so this name fit the case now. But, in a future, if we probably add a new interface for the listing (e.g., `SparkSession.analyzeTable`, `SparkSession.analyzeColumn`, ...), this name a little confuses users? So, how about `spark.sql.parallelFileListingInStatsComputation.enabledl`? @gatorsmile

File: docs/sql-programming-guide.md (hunk 18 above)
Comment (Achuth17) [reply docs/sql-programming-guide.md]: @maropu The calculateTotalSize flow is invoked in other places too. For eg, it is used in updateTableStats method which is called from a few places.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 19 above)
Comment (Achuth17) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I can make this change..

File: docs/sql-programming-guide.md (hunk 17 above)
Comment (Achuth17) [reply docs/sql-programming-guide.md]: Same as below.

File: docs/sql-programming-guide.md (hunk 18 above)
Comment (maropu) [reply docs/sql-programming-guide.md]: ah, ok.

**Comment by Achuth17** [comment]
@maropu @gatorsmile The tests are failing in places unrelated to my changes. How can I resolve this?

**Comment by Achuth17** [comment]
retest this please.

**Comment by maropu** [comment]
retest this please

**Comment by Achuth17** [comment]
@maropu @gatorsmile, Any other changes required?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 20]
```diff
@@ -49,4 +51,11 @@ object DataSourceUtils {
       }
     }
   }
+
+  // SPARK-15895: Metadata files (e.g. Parquet summary files) and temporary files should not be
+  // counted as data files, so that they shouldn't participate partition discovery.
+  private[sql] def isDataPath(path: Path): Boolean = {
+    val name = path.getName
+    !((name.startsWith("_") && !name.contains("=")) || name.startsWith("."))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala]: Let us do not use the same one with `PartitioningAwareFileIndex.scala`. In this case, the data file is not related to the following condition `name.contains("=")`

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala]: Should I use the earlier implementation with a simple if condition?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala]: Not sure what is your earlier impl. I would prefer to keeping unchanged the original code in `PartitioningAwareFileIndex.scala`. Just add a utility function `isDataPath ` in CommandUtils.scala. Does this sound good to you?

Comment (Achuth17) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala]: Sounds good. I have made the changes.

**Comment by gatorsmile** [comment]
Thanks! Merged to master.
